File: crates/belfast/src/compute.rs
use crate::{BelfastError, BelfastResult, BindGroup, Device};

pub struct ComputeOptions<'a> {
    pub label: &'a str,
    pub layout: Option<&'a wgpu::PipelineLayout>,
    pub entry_point: &'a str,
}

impl<'a> ComputeOptions<'a> {
    pub fn new(label: &'a str) -> Self {
        Self {
            label,
            layout: None,
            entry_point: "cs_main",
        }
    }
}

pub struct Compute {
    pipeline: wgpu::ComputePipeline,
    device: Device,
}

impl Compute {
    pub fn new(device: &Device, shader_code: &str, options: ComputeOptions<'_>) -> Self {
        let module = device
            .gpu()
            .create_shader_module(wgpu::ShaderModuleDescriptor {
                label: Some(&format!("{}Shader", options.label)),
                source: wgpu::ShaderSource::Wgsl(shader_code.into()),
            });
        let pipeline = device
            .gpu()
            .create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
                label: Some(&format!("{}Pipeline", options.label)),
                layout: options.layout,
                module: &module,
                entry_point: Some(options.entry_point),
                compilation_options: Default::default(),
                cache: None,
            });

        Self {
            pipeline,
            device: device.clone(),
        }
    }

    pub fn get_bind_group_layout(&self, index: u32) -> wgpu::BindGroupLayout {
        self.pipeline.get_bind_group_layout(index)
    }

    pub fn device(&self) -> &Device {
        &self.device
    }

    pub fn validate_for_dispatch(&self, device: &Device) -> BelfastResult<()> {
        if !self.device.is_same(device) {
            return Err(BelfastError::ComputeDeviceMismatch);
        }
        Ok(())
    }

    pub fn dispatch<'a>(&'a self, pass: &mut wgpu::ComputePass<'a>, workgroups: [u32; 3]) {
        pass.set_pipeline(&self.pipeline);
        pass.dispatch_workgroups(workgroups[0], workgroups[1], workgroups[2]);
    }

    pub fn run(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        bind_group: Option<&BindGroup>,
        workgroups: [u32; 3],
        label: Option<&str>,
    ) {
        let mut pass = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
            label,
            timestamp_writes: None,
        });
        if let Some(bind_group) = bind_group {
            bind_group.bind_compute(&mut pass, 0);
        }
        self.dispatch(&mut pass, workgroups);
    }
}

File: archive/packages/belfast/src/helper/Compute.ts
import type { BindGroup } from "../core/BindGroup";
import type { Device } from "../core/Device";
import { createComputePipeline, createShaderModule } from "../core/GPUResources";

export interface ComputeOptions {
  label?: string;
  /** Defaults to `"auto"`. Use a shared layout to reuse bind groups across pipelines. */
  layout?: GPUPipelineLayout | "auto";
  /** Compute entry point. Defaults to `"cs_main"`. */
  entryPoint?: string;
}

/** Number of workgroups to dispatch: `x`, `[x, y]`, or `[x, y, z]`. */
export type WorkgroupCount = number | readonly [number, number] | readonly [number, number, number];

/**
 * Compute pipeline wrapper mirroring {@link Draw}.
 * Encapsulates pipeline creation and reduces a compute dispatch to one call.
 */
export class Compute {
  private pipeline: GPUComputePipeline;

  constructor(device: Device, shaderCode: string, optionsOrLabel: ComputeOptions | string = {}) {
    const options = typeof optionsOrLabel === "string" ? { label: optionsOrLabel } : optionsOrLabel;
    const { label = "Compute", layout = "auto", entryPoint = "cs_main" } = options;

    const module = createShaderModule(device, shaderCode, `${label}Shader`);

    this.pipeline = createComputePipeline(device, {
      label: `${label}Pipeline`,
      layout,
      compute: { module, entryPoint },
    });
  }

  getBindGroupLayout(index = 0): GPUBindGroupLayout {
    return this.pipeline.getBindGroupLayout(index);
  }

  /** Records pipeline + bind groups + dispatch into an existing compute pass. */
  dispatch(
    passEncoder: GPUComputePassEncoder,
    bindGroup?: BindGroup | readonly BindGroup[],
    workgroups: WorkgroupCount = 1,
  ): void {
    passEncoder.setPipeline(this.pipeline);
    if (bindGroup) {
      const groups = Array.isArray(bindGroup) ? bindGroup : [bindGroup];
      for (let i = 0; i < groups.length; i++) {
        passEncoder.setBindGroup(i, groups[i].gpu);
      }
    }
    if (typeof workgroups === "number") {
      passEncoder.dispatchWorkgroups(workgroups);
    } else {
      passEncoder.dispatchWorkgroups(workgroups[0], workgroups[1] ?? 1, workgroups[2] ?? 1);
    }
  }

  /** Convenience: begin a compute pass, dispatch, and end it in a single call. */
  run(
    encoder: GPUCommandEncoder,
    bindGroup?: BindGroup | readonly BindGroup[],
    workgroups: WorkgroupCount = 1,
    label?: string,
  ): void {
    const pass = encoder.beginComputePass(label ? { label } : undefined);
    this.dispatch(pass, bindGroup, workgroups);
    pass.end();
  }
}
